Expansion of 'turnbattle' system contrib: tb_range

FlutterSprite opened the pull request:
Adds a new contrib module: tb_range, an expansion for the turnbattle contrib which implements a system of movement and positioning.

#### Brief overview of PR changes/additions
The tb_range module includes an abstract system of movement and position in combat, which works by tracking different fighters and objects' relative distance to each other without tracking their absolute position. This allows for a simple and intuitive (if imprecise) system for movement in a text-based game that doesn't disadvantage players on screen readers or require navigating around a grid.

#### Motivation for adding to Evennia
I plan to add a few more expansions to the 'turnbattle' contrib folder demonstrating various expansions to the basic combat systems, and then a 'complete' version which implements all of the subsystems of the various modules and implements interactions between them (like melee and ranged weapons for the equipment system in tb_equip, for example). Ultimately, the goal is to have a fully-featured combat system ready for aspiring MUD creators who want to get their game off the ground right away, and to give many options to let them use as much or as little of it as they like.

#### Other info (issues closed, discussion etc)
This system, like the 'turnbattle' contrib itself, is based around its implementation in my learning project "The World of Cool Battles", though it has been streamlined in some places (fewer possible range values and lack of multi-step movement) and expanded in others (the tracking of distances to objects and exits).

FlutterSprite commented:
The unit tests aren't in yet, but I'd like to know what everyone thinks! This contrib's a little more complex than my others.

Griatch commented on evennia/contrib/turnbattle/tb_range.py:
```diff
@@ -0,0 +1,1356 @@
+        if defender.db.hp <= 0:
+            attacker.location.msg_contents("%s has been defeated!" % defender)
+            
+def distance_dec(mover, target):
```
From how this is used in-code, `distance_dec` and `distance_inc` are helper functions to `approach` and `withdraw`, the latter being the functions intended to be used externally. If this is the case you should clarify which functions are part of your API and which are internal-use only. This can be done either by marking these helper-functions as private (`_distance_dec/inc`) or by embedding them inside their main functions.

FlutterSprite replied:
Ah! This was something I knew nothing about - I'll definitely address this. There's not really anywhere else I would think to use distance_dec and distance_inc by themselves.

Griatch commented on evennia/contrib/turnbattle/tb_range.py:
```diff
@@ -0,0 +1,1356 @@
+        target.db.Combat_Range[mover] = 0
+        mover.db.Combat_Range = target.db.Combat_Range
+        # Assure everything else has the same distance from the mover and target, now that they're together
+        for object in mover.location.contents:
```
`object` is a reserved Python word. Never use a reserved word as a variable name even if you get no error from it here  - not only is it a potential source of insidious bugs, it's also against PEP8 and Python praxis and will also trigger all sorts of warnings in external style checkers down the line. Use something like `obj` instead.

FlutterSprite replied:
Whoops!! I'll definitely fix this.

Griatch commented on evennia/contrib/turnbattle/tb_range.py:
```diff
@@ -0,0 +1,1356 @@
+                distance_inc(mover, thing)
+    # Lastly, move closer to your target.
+    distance_dec(mover, target)
+    return 
```
No need for the empty `return` here; it's implicit.

Griatch commented on evennia/contrib/turnbattle/tb_range.py:
```diff
@@ -0,0 +1,1356 @@
+                distance_inc(mover, thing)
+    # Then, move away from your target.
+    distance_inc(mover, target)
+    return
```
Empty `return` has no function here.

Griatch commented on evennia/contrib/turnbattle/tb_range.py:
```diff
@@ -0,0 +1,1356 @@
+        range (int or None): Distance between two objects or None if not applicable
+    """
+    # Return None if not applicable.
+    if not obj1.db.Combat_Range:
```
Out of curiosity, there is no validation of `db.Combat_Range` in `distance_inc/dec` nor in `approach/withdraw`. If there is a risk of using objects without such an Attribute the check should probably either be extended to those too, or removed here for consistency (that is, if you use objects not valid for combat you should always expect a traceback).

FlutterSprite replied:
I do think this is a good function to have, though it was intuition more than anything rational that lead me to make it instead of referencing db.combat_range directly - I'll add it into those functions too.

Griatch commented on evennia/contrib/turnbattle/tb_range.py:
```diff
@@ -0,0 +1,1356 @@
+    Notes:
+        Any attribute whose key begins with 'combat_' is temporary and no
+        longer needed once a fight ends.
+    """
```
It's an interesting take to use the combattants themselves to store range combat data rather than to track things on the combat handler. :)

FlutterSprite replied:
Since combatants don't have absolute positions, only relative distances to each other, this seemed the most intuitive way to do it!

Griatch commented on evennia/contrib/turnbattle/tb_range.py:
```diff
@@ -0,0 +1,1356 @@
+    Returns:
+        (bool): True if in combat or False if not in combat
+    """
+    if character.db.Combat_TurnHandler:
```
This can simply and more succinctly be expressed as `return bool(character.db.Combat_TurnHandler)`.

Griatch commented on evennia/contrib/turnbattle/tb_range.py:
```diff
@@ -0,0 +1,1356 @@
+    turnhandler = character.db.Combat_TurnHandler
+    currentchar = turnhandler.db.fighters[turnhandler.db.turn]
+    if character == currentchar:
+        return True
```
See above.

Griatch commented on evennia/contrib/turnbattle/tb_range.py:
```diff
@@ -0,0 +1,1356 @@
+        if is_in_combat(dropper) and not is_turn(dropper):
+            dropper.msg("You can only drop things on your turn!")
+            return False
+        return True
```
Minor PEP8 nag: Put one empty line between all methods in a class.

Griatch commented on evennia/contrib/turnbattle/tb_range.py:
```diff
@@ -0,0 +1,1356 @@
+"""
+
+
+class TBRangeTurnHandler(DefaultScript):
```
I am a little divided on placing this important component at the end of the module. The handler is used and referenced  throughout the other components, so when reading the code one wonders what it is and where it comes from. For this reason I always push for a module to be ordered so that things are defined above the code that use them. Hence I feel this should be defined higher up, or even at the top.

FlutterSprite replied:
I think it was my thought originally that the turn handler used lots of the other functions, it should come after them, but at some point I moved many of those functions into the turn handler itself. I'll put it higher up here and in the rest of the modules.

Griatch commented on evennia/contrib/turnbattle/tb_range.py:
```diff
@@ -0,0 +1,1356 @@
+        if self.db.turn > len(self.db.fighters) - 1:
+            self.db.turn = 0  # Go back to the first in the turn order once you reach the end.
+        newchar = self.db.fighters[self.db.turn]  # Note the new character
+        self.db.timer = 30 + self.time_until_next_repeat()  # Reset the timer.
```
Timeouts like these seem very game-specific to hide away inside the Script body like this. Might it not be an idea to have a block of global constants up top, like `TURN_TIMEOUT = 30` for things like this? Would also make it easier for the dev to see which knobs are easy to tweak ...

FlutterSprite replied:
This is definitely a good idea - actions per turn would be another good variable to be able to easily mess with. I think it was my original thought that adding extra flexibility with variables would complicate things too much for a system that's mostly meant to be completely overhauled, but some things like this would be universal enough that I can't imagine they'd be taken out completely!

FlutterSprite commented on evennia/contrib/turnbattle/tb_range.py:
```diff
@@ -0,0 +1,1356 @@
+        attacker.location.msg_contents("%s hits %s with a %s attack for %i damage!" % (attacker, defender, attack_type, damage_value))
+        apply_damage(defender, damage_value)
+        # If defender HP is reduced to 0 or less, announce defeat.
+        if defender.db.hp <= 0:
```
This would be a good thing to separate out - I'll have to apply this change across all the 'turnbattle' modules.

FlutterSprite commented on evennia/contrib/turnbattle/tb_range.py:
```diff
@@ -0,0 +1,1356 @@
+        mover (obj): The object moving
+        target (obj): The object to be moved toward
+    """
+    mover.db.Combat_Range[target] -= 1
```
This would be an easy change! I'll apply this across the rest of the modules too.

FlutterSprite commented:
There! Still no unit tests, but I addressed all the concerns you raised. This involved going back and making changes in tb_basic and tb_equip as well.

Griatch commented:
Cool! I'll review again when the unit tests are in then. :)

FlutterSprite commented:
The unit tests are in! This one was quite a pain.

Griatch commented:
@FlutterSprite This looks good I think. There is one thing I'm missing here though, when reading the contrib's documentation, and that is how decoupled these various modules are. That is, when I look at the `README.md` I want to know if I can combine equiplent + range, if I need basic to use range ... and so on. So while the actual install instructions are in each file, I think it would be enlightening with some more info about these things, as well as pointing out where the install instructions are (since those are normally in the `README.md` file for other contribs).

FlutterSprite commented:
Gotcha! I'll clarify things in the readme real quick.

FlutterSprite commented:
How's that?

Griatch commented:
Thanks, merged now! One thing to remember for the next time - for your own sanity of mind, don't develop things like this in your own copy of the upstream develop (or master) branch. It's better that you keep your version of develop in sync with upstream and do your development in a new branch off that (this you can then just delete once your contributions is available upstream). Otherwise, if I should for example rebase or squash your PR's commits, your version of devel/master branch will be different from the upstream one and you'll face merge conflicts.

FlutterSprite commented:
Oh, that makes a lot of sense! Slowly but surely, I'm learning how to use GitHub properly. This is the first time I've actually committed my changes from the command line instead of pasting them into the website like a scrub.
